akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ClusterSharding.scala

```diff
@@ -195,68 +246,112 @@ final class AdaptedClusterShardingImpl(system: ActorSystem[_]) extends ClusterSh
     handOffStopMessage: A): ActorRef[ShardingEnvelope[A]] = {
-    val extractor = new HashCodeMessageExtractor[A](maxNumberOfShards)
-    spawn(behavior, entityProps, typeKey, settings, extractor, defaultShardAllocationStrategy(settings), handOffStopMessage)
+    val extractor = new HashCodeMessageExtractor[A](maxNumberOfShards, handOffStopMessage)
+    spawn2(behavior, entityProps, typeKey, settings, extractor, defaultShardAllocationStrategy(settings))
   }
 
-  override def spawn[E, A](
-    behavior:           Behavior[A],
+  override def spawnJavadsl[A](
+    behavior:           EntityIdToBehavior[A],
     entityProps:        Props,
     typeKey:            EntityTypeKey[A],
     settings:           ClusterShardingSettings,
-    messageExtractor:   ShardingMessageExtractor[E, A],
-    handOffStopMessage: A): ActorRef[E] =
-    spawn(behavior, entityProps, typeKey, settings, messageExtractor, defaultShardAllocationStrategy(settings), handOffStopMessage)
+    maxNumberOfShards:  Int,
+    handOffStopMessage: A): ActorRef[ShardingEnvelope[A]] = {
+    val extractor = new HashCodeMessageExtractor[A](maxNumberOfShards, handOffStopMessage)
+    spawnJavadsl(behavior, entityProps, typeKey, settings, extractor, defaultShardAllocationStrategy(settings))
+  }
 
-  override def spawn[E, A](
-    behavior:           Behavior[A],
+  override def spawn3[E, A](
+    behavior:         String ⇒ Behavior[A],
+    entityProps:      Props,
+    typeKey:          EntityTypeKey[A],
+    settings:         ClusterShardingSettings,
+    messageExtractor: ShardingMessageExtractor[E, A]): ActorRef[E] =
+    spawn2(behavior, entityProps, typeKey, settings, messageExtractor, defaultShardAllocationStrategy(settings))
+
+  override def spawnJavadsl[E, A](
+    behavior:         EntityIdToBehavior[A],
+    entityProps:      Props,
+    typeKey:          EntityTypeKey[A],
+    settings:         ClusterShardingSettings,
+    messageExtractor: ShardingMessageExtractor[E, A]): ActorRef[E] =
+    spawnJavadsl(behavior, entityProps, typeKey, settings, messageExtractor, defaultShardAllocationStrategy(settings))
+
+  override def spawn2[E, A](
+    behavior:           String ⇒ Behavior[A],
     entityProps:        Props,
     typeKey:            EntityTypeKey[A],
     settings:           ClusterShardingSettings,
     extractor:          ShardingMessageExtractor[E, A],
-    allocationStrategy: ShardAllocationStrategy,
-    handOffStopMessage: A): ActorRef[E] = {
+    allocationStrategy: ShardAllocationStrategy): ActorRef[E] = {
 
     val untypedSettings = ClusterShardingSettings.toUntypedSettings(settings)
 
+    val extractorAdapter = new ExtractorAdapter(extractor)
+    val extractEntityId: ShardRegion.ExtractEntityId = {
+      // TODO is it possible to avoid the double evaluation of entityId
+      case message if extractorAdapter.entityId(message) != null ⇒
+        (extractorAdapter.entityId(message), extractorAdapter.unwrapMessage(message))
+    }
```

Comment:
Um, so match error on null here? Why not just:
```
val id = extractorAdapter.entityId(message)
if (id != null) (id, extractorAdapter.unwrapMessage(message))
else throw IllegalSomethingException(s"Adapter returned null for message [$message.getClass]")
```

Reply:
it's a partial function that I wanted to play nice with